File: java/generateparentheses.java
public class Solution {
    
    private ArrayList<String> ans;
    private int n;
    
    private void gP(int left, int right, String tmp){
        if(left<right) return;
        if(left>n) return;
        if(right>n) return;
        
        if(left==n && right==n) ans.add(tmp);
        String tmp1 = tmp+"(";
        String tmp2 = tmp+")";
        gP(left+1,right,tmp1);
        gP(left,right+1,tmp2);
    }
    
    public ArrayList<String> generateParenthesis(int n) {
        // Start typing your Java solution below
        // DO NOT write main() function
        ans = new ArrayList<String>();
        this.n = n;
        gP(0,0,"");
        return ans;
    }
}

File: cpp/GenerateParentheses.cpp
class Solution {
public:
    
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        generateP(0,0,res,"", n);
        return res;
    }
    
    void generateP(int left, int right, vector<string>& res, string s, int n) {
        if(left==n && right==n) {
            res.push_back(s);
            return;
        }
        if(left>right) {
            if(right+1<=n) {
                generateP(left, right+1, res, s+')',n);
            }
            if(left+1<=n) {
                generateP(left+1, right, res, s+'(',n);
            }
        }
        if(left==right && left+1<=n) {
            generateP(left+1, right, res, s+'(',n);
        }
    }
};
